core/sail/sail-spin/src/test/java/org/eclipse/rdf4j/sail/spin/SpinSailTest.java

```diff
@@ -141,6 +147,39 @@ public void testSpinxRule() throws Exception {
 		assertStatements("testSpinxRule-expected.ttl");
 	}
 
+	@Ignore("This test shows how the SpinSail fails on transactional workloads.")
+	@Test
+	public void testDeepConstructRule() throws Exception {
+		loadStatements("testDeepConstructRule.ttl");
+
+		ValueFactory vf = SimpleValueFactory.getInstance();
+
+		String ns = "http://example.org/";
+
+		IRI ex1 = vf.createIRI(ns, "ex1");
+		IRI TestClass = vf.createIRI(ns, "TestClass");
+		IRI prop = vf.createIRI(ns, "prop");
+		IRI ex2 = vf.createIRI(ns, "ex2");
+		IRI ex3 = vf.createIRI(ns, "ex3");
+		IRI ex4 = vf.createIRI(ns, "ex4");
+
+		conn.begin();
+		conn.add(ex1, RDF.TYPE, TestClass);
+		conn.add(ex1, prop, ex2);
+
+		// comment out these two lines, so that all statements are added in the same transaction. This will make the
+		// test pass.
+		conn.commit();
+		conn.begin();
+
+		conn.add(ex2, FOAF.KNOWS, ex3);
+		conn.add(ex3, FOAF.KNOWS, ex4);
+		conn.commit();
+
+		assertStatements("testDeepConstructRule-expected.ttl");
+
+	}
```

Comment:
@jeenbroekstra 
I've added this test here to show the lack of transactional support in the SpinSail. Is there some way for us to make this kind of scenario function? Eg. by having the SpinSail always recompute every rule on every transaction?

Reply:
Essentially, the current implementation only triggers a computation when a matching type statement is added. So if it is added in the first transaction, and later we add more data that should actually trigger the rule....then nothing is inferred.

Reply:
It's actually a little bit better, it's triggered when a resource is modified if that resource has a matching type.